dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -627,25 +661,66 @@ static void AddResponseMessage(ChatCompletionsOptions chatOptions, ChatHistory c
             string? streamedName = null;
             ChatRole? streamedRole = default;
             CompletionsFinishReason finishReason = default;
-            await foreach (StreamingChatCompletionsUpdate update in response.ConfigureAwait(false))
+
+            using (var activity = ModelDiagnostics.StartCompletionActivity(this.Endpoint, this.DeploymentOrModelName, ModelProvider, chat, executionSettings))
             {
-                metadata = GetResponseMetadata(update);
-                streamedRole ??= update.Role;
-                streamedName ??= update.AuthorName;
-                finishReason = update.FinishReason ?? default;
+                // Make the request.
+                StreamingResponse<StreamingChatCompletionsUpdate> response;
+                try
+                {
+                    response = await RunRequestAsync(() => this.Client.GetChatCompletionsStreamingAsync(chatOptions, cancellationToken)).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    activity?.SetError(ex);
+                    throw;
+                }
 
-                // If we're intending to invoke function calls, we need to consume that function call information.
-                if (autoInvoke)
+                var responseEnumerator = response.ConfigureAwait(false).GetAsyncEnumerator();
+                List<OpenAIStreamingChatMessageContent> streamedContents = [];
```

Comment:
It'd be nice if we only paid to store all these items if diagnostics are enabled, e.g. change this to be `activity is not null ? [] : null` and then the Adds to be more like `streamedContents?.Add(openAIStreamingChatMessageContent);`.